### ISSUE TITLE
Request delegate generator produces CS8073 for ValueTask<T?> results

### ISSUE THREAD

**Issue by gfoidl** [dotnet/aspnetcore#54172]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

For delegates that return `ValueTask<T?>? the compiler emits
```
Warning	CS8073	The result of the expression is always 'false' since a value of type 'ValueTask<Person>' is never equal to 'null' of type 'ValueTask<Person>?'
```

This is especially cumbersome when 
```xml
<TreatWarningsAsErrors>true</TreatWarningsAsErrors>
```
is set.

### Expected Behavior

No warning is produced.

### Steps To Reproduce

```c#
WebApplicationBuilder builder = WebApplication.CreateBuilder(args);
WebApplication app = builder.Build();

Person[] persons = [new Person("Batman"), new Person("Me")];

// No problem
Person? SearchPerson0(string name) => Array.Find(persons, p => p.Name == name);

// Here the ValueTask causes the CS8073
ValueTask<Person?> SearchPerson1(string name)
{
    Person? person = Array.Find(persons, p => p.Name == name);
    return ValueTask.FromResult(person);
}

// This is fine, as Task is a reference type
Task<Person?> SearchPerson2(string name)
{
    Person? person = Array.Find(persons, p => p.Name == name);
    return Task.FromResult(person);
}

app.MapGet("person0/{name}", SearchPerson0);
app.MapGet("person1/{name}", SearchPerson1);
app.MapGet("person2/{name}", SearchPerson2);

app.Run();

public record Person(string Name);
```

```xml
<Project Sdk="Microsoft.NET.Sdk.Web">

    <PropertyGroup>
        <TargetFramework>net8.0</TargetFramework>
        <Nullable>enable</Nullable>
        <ImplicitUsings>enable</ImplicitUsings>
    </PropertyGroup>

    <PropertyGroup>
        <IncludeOpenAPIAnalyzers>true</IncludeOpenAPIAnalyzers>

        <EnableRequestDelegateGenerator>true</EnableRequestDelegateGenerator>
        <InterceptorsPreviewNamespaces>$(InterceptorsPreviewNamespaces);Microsoft.AspNetCore.Http.Generated</InterceptorsPreviewNamespaces>
    </PropertyGroup>

</Project>
```

### .NET Version

8.0.200

### Anything else?

If this got already fixed for .NET 9 is there a chance to backport it to .NET 8 as `<TreatWarningsAsErrors>` needs some workarounds to make it compile and .NET 8 is a LTS?

**Comment by captainsafia** [comment]
@gfoidl Thanks for reporting this issue! It still repros in .NET 9 so there's some work to do here.

Would you be interested in submitting a PR to resolve this issue?

**Comment by tcortega** [comment]
I'd like to take this one if that's ok @captainsafia

**Comment by gfoidl** [comment]
@tcortega great 👍🏻 I assigned the issue to you (and regarding @captainsafia's question about a PR from me: I don't have time for this the next days, so I'm happy you can work on this).

**Comment by tcortega** [comment]
> @tcortega great 👍🏻 I assigned the issue to you (and regarding @captainsafia's question about a PR from me: I don't have time for this the next days, so I'm happy you can work on this).

Oh man, my bad, I thought that @captainsafia was asking for anyone, I misread her question. Thanks tho.

**Comment by tcortega** [comment]
I've understood where the problem is and why it occurs. It seems there's a bug in the response type check itself, where when the response type is annotated as nullable, the code ends up checking if the Task is null, as seen here:
https://github.com/dotnet/aspnetcore/blob/c7c76d7d3bb30e29e8386d1a3e6dae9e307fd0de/src/Http/Http.Extensions/gen/StaticRouteHandlerModel/StaticRouteHandlerModel.Emitter.cs#L91-L97

And here:
https://github.com/dotnet/aspnetcore/blob/c7c76d7d3bb30e29e8386d1a3e6dae9e307fd0de/src/Http/Http.Extensions/gen/StaticRouteHandlerModel/StaticRouteHandlerModel.Emitter.cs#L383-L389

The problem is that the IF check is incorrect. When it checks: `endpoint.Response?.ResponseType?.NullableAnnotation`, what it actually checks is whether the inner type of the `awaitable type` is nullable or not. If the type were, for instance, `Task<bool>?`, it would never fall into this check because `bool` was not annotated as `nullable`.

With that said, I believe the correct procedure would be to check from the awaitable type itself; however, to do this, I would have to modify how the [EndpointResponse](https://github.com/dotnet/aspnetcore/blob/main/src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointResponse.cs) class works.

So that this line:
https://github.com/dotnet/aspnetcore/blob/c7c76d7d3bb30e29e8386d1a3e6dae9e307fd0de/src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointResponse.cs#L30

Stores the type itself, and not the string.

I would like to hear your opinions on this @captainsafia and @gfoidl, and if everything is okay, I plan to open the PR.

**Comment by gfoidl** [comment]
Yep, the type itself should be stored, so that it can be checked whether it's a valuetype or a referencetype later on.
Just as you said.

**Comment by captainsafia** [comment]
> Oh man, my bad, I thought that @captainsafia was asking for anyone, I misread her question. Thanks tho.

Yep, it wasn't clear in my response but it was intended as a invite for contribution for anyone. Hence the https://github.com/dotnet/aspnetcore/labels/help%20wanted label, so thanks for taking it on!

> Stores the type itself, and not the string.

This proposal sounds good to me. Thanks for looking into the issue!

**Comment by tcortega** [comment]
Thanks for the feedback, the PR is already opened! #54335
